### DISCUSSION TITLE
Sample latents from VAE to generate close images

### DISCUSSION THREAD

**Discussion by Clement-Lelievre** [huggingface/diffusers#9327] [Q&A]
hi,

I would like to use a VAE (SDXL's VAE in the example below) in order to get close variations of a given image, in the spirit of what was demonstrated in [that paper](https://arxiv.org/abs/1609.04468).
I could achieve this with inpainting or maybe img2img, but my point is to get these close images without having to go through reverse diffusion, for performance reasons.

I tried to implement the logic, and found out that diffusers already has nearly all the tools.

**PROBLEM**: It does run without errors and I do get images that are different pixel-wise to the initial image, however the difference across images is invisible to the human eye. I noticed that the std of the latent distribution is extremely low relative to the mean, that might explain why all the produced images look so identical to the input? I might have done a preproc step wrong also?

I'd be grateful if someone could review my logic and advise.
Here is my code:

```python
import torch
from diffusers import AutoencoderKL
from diffusers.image_processor import VaeImageProcessor
from diffusers.models.autoencoders.vae import DiagonalGaussianDistribution
from PIL import Image
from torchvision import transforms
from torchvision.transforms.functional import pil_to_tensor

# 2: assume I'm interested in getting close variations of an image, using SDXL's VAE
# instead of inferring, eg doing an img2img or inpaint,
# I can use the latent proba distribution inferred by the VAE, sample from it and decode back to pixel space

# instantiate SDXL's VAE
with torch.no_grad():
    # vae:AutoencoderKL = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix")
    vae: AutoencoderKL = AutoencoderKL.from_pretrained(
        "stabilityai/stable-diffusion-xl-base-1.0", subfolder="vae"
    )

    vae.to(dtype=torch.float32)  # otherwise it produces NaNs, even madebyollin's VAE
    vae.to(device="cuda")

    assert vae.device == torch.device("cuda:0")
    assert vae.dtype == torch.float32

    # make image as tensor
    img = Image.open(<image_path_on_disk>)
    img_tensor = pil_to_tensor(img).unsqueeze(0) / 255.0
    img_tensor = img_tensor.to(vae.device)
    img_tensor = img_tensor.to(vae.dtype)

    # get the inferred latent distribution
    latent_dist: DiagonalGaussianDistribution = vae.encode(img_tensor, return_dict=False)[0]
    print(
        f"{latent_dist.mean.shape=} {latent_dist.std.shape=} {latent_dist.mean.mean()=} {latent_dist.std.mean()=}"
    )
    assert not latent_dist.mean.isnan().any()
    assert not latent_dist.std.isnan().any()
    assert latent_dist.deterministic is False

    # generate another latents from it
    sample_1 = (
        latent_dist.sample()
    )  # take the mean of the latent dist and add a fraction of the std
    sample_2 = latent_dist.sample()
    assert not sample_1.isnan().any()
    assert not sample_2.isnan().any()
    assert (sample_1 != sample_2).any(), "samples should be different"

    print(f"{sample_1.shape=}")

    # sample_1 *= (2 ** (len(vae.config.block_out_channels) - 1))

    assert vae.dtype == sample_1.dtype
    assert vae.device == sample_1.device

    img_1: torch.Tensor = vae.decode(sample_1).sample  # , return_dict=False)[0]
    ##vae_scale_factor = 2 ** (len(vae.config.block_out_channels) - 1)
    img_1 = img_1.squeeze(0).cpu().detach()
    assert (
        img_1 != img_tensor.cpu().detach()
    ).any(), "generated image should be different from the input image"
    img_1_pil = transforms.ToPILImage()(img_1)
    # vae_img_processor = VaeImageProcessor(vae_scale_factor=2 ** (len(vae.config.block_out_channels) - 1))
    # img_1_pil = vae_img_processor.postprocess(img_1)[0]
    img_1_pil.save("sample_1.png")
```

Thanks a lot!

cc @asomoza

**Comment by bmaxdk** [comment]
I tested with your code with updating noise and scale factor with some modification.

```py
import torch
from diffusers import AutoencoderKL
from diffusers.image_processor import VaeImageProcessor
from diffusers.models.autoencoders.vae import DiagonalGaussianDistribution
from PIL import Image
from torchvision import transforms
from torchvision.transforms.functional import pil_to_tensor

# 2: assume I'm interested in getting close variations of an image, using SDXL's VAE
# instead of inferring, eg doing an img2img or inpaint,
# I can use the latent proba distribution inferred by the VAE, sample from it and decode back to pixel space

# Instantiate SDXL's VAE
with torch.no_grad():
    # vae:AutoencoderKL = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix")
    vae: AutoencoderKL = AutoencoderKL.from_pretrained(
        "stabilityai/stable-diffusion-xl-base-1.0", subfolder="vae"
    )

    vae.to(dtype=torch.float32) # otherwise it produces NaNs, even madebyollin's VAE
    vae.to(device="cuda")

    assert vae.device == torch.device("cuda:0")
    assert vae.dtype == torch.float32

    # make image as tensor
    img = Image.open("avenger.jpg")  # Replace with your actual image path
    img_tensor = pil_to_tensor(img).unsqueeze(0) / 255.0
    img_tensor = img_tensor.to(vae.device)
    img_tensor = img_tensor.to(vae.dtype)

    # get the inferred latent distribution
    latent_dist: DiagonalGaussianDistribution = vae.encode(img_tensor, return_dict=False)[0]
    print(f"{latent_dist.mean.shape=} {latent_dist.std.shape=} {latent_dist.mean.mean()=} {latent_dist.std.mean()=}")

    assert not latent_dist.mean.isnan().any()
    assert not latent_dist.std.isnan().any()
    assert latent_dist.deterministic is False

    # -- Tried with scale factor and add noise--
    scale_factor = 5.0  # increased the variations
    noise_strength = 0.2  # add noise will help further perturb the latent space
    noise = noise_strength * torch.randn_like(latent_dist.mean).to(vae.device)

    # generate new latents with added noise and scaling
    sample_1 = latent_dist.mean + scale_factor * latent_dist.std * torch.randn_like(latent_dist.mean).to(vae.device) + noise
    sample_2 = latent_dist.mean + scale_factor * latent_dist.std * torch.randn_like(latent_dist.mean).to(vae.device) + noise
    assert not sample_1.isnan().any()
    assert not sample_2.isnan().any()
    assert (sample_1 != sample_2).any(), "samples should be different"

    print(f"{sample_1.shape=}")

    assert vae.dtype == sample_1.dtype
    assert vae.device == sample_1.device

    # decode the sampled latents back to images
    img_1: torch.Tensor = vae.decode(sample_1).sample  # Decoding the first variation
    img_1 = img_1.squeeze(0).cpu().detach()
    assert (img_1 != img_tensor.cpu().detach()).any(), "generated image should be different from the input image"
    
    # save first variation
    img_1_pil = transforms.ToPILImage()(img_1)
    img_1_pil.save("sample2_1.png")

    
    # save second variation 
    img_2: torch.Tensor = vae.decode(sample_2).sample
    img_2 = img_2.squeeze(0).cpu().detach()
    img_2_pil = transforms.ToPILImage()(img_2)
    img_2_pil.save("sample2_2.png")
    
    # -- Try interpolation between Latent vectors --
    t = torch.rand(1).item()  # generates a random interpolation factor between 0 - 1
    interpolated_sample = (1 - t) * sample_1 + t * sample_2
    
    # Decode the interpolated sample --> image
    img_interpolated = vae.decode(interpolated_sample).sample
    img_interpolated = img_interpolated.squeeze(0).cpu().detach()
    img_interpolated = (img_interpolated * 0.5 + 0.5).clamp(0, 1)
    img_interpolated_pil = transforms.ToPILImage()(img_interpolated)
    img_interpolated_pil.save("interpolated_variation2.png")
    print("Done")


```

It sure does working well. It does change eyes, smiles, etc. You may want to see significant changes. However, you may need to consider the dimension of the latent space. Also VAE are good at generating smooth variation. it may not be as effective at huge change unless it trained for that.

**Reply by Clement-Lelievre** [reply]
hi @bmaxdk 

thanks for your reply.

Could you please share an input image or several, that worked on your side, the intermediate images (sample2_1.png and sample2_2.png) and the variations (interpolated_variation2.png) you get?
I'm interested in seeing the size and verify that it's RGB.

I tried with various sizes, eg 1024x1024 and then with relatively small images like 32x32 in order for the latent space to be much smaller and to have more impact on it.

Two issues on my side when running your snippet:
- the two samples are very close to one another, likely because the latent distributions std are very small (in the order of 10^-5) and dominate the random tensors; so the interpolation does not really play a role IMO, at least in the examples I ran. When increasing the scale_factor to match this magnitude, the decoded samples are indeed different but too degraded
- the final image (interpolated_variation2.png) looks a washed out version of the input image on my end (this is eliminated when avoiding the `img_interpolated * 0.5 + 0.5` part )

Notes:
- I tried with SD1.5's VAE and it seemed like the latent dimensionality is the same as SDXL's VAE
- I tried interpolating between an image of a non-smiling portrait and its smiling equivalent, and decoding the averaged latents produced a half smiling portrait. But I did pass the two images as inputs, these are not the two latents from your code. I'd have to confirm with more images but seems like interpolation worked. This thus requires two input images and they need to be close and the difference needs to be "continuous" I think, not binary (eg smiling/not smiling intead of glasses/no glasses).

However I'm still interested in getting to see your version of my original snippet work, as this would not require two input images+the above conditions, but just one.

**Comment by Clement-Lelievre** [comment]
Here is an exmaple of what I get when interpolating two images (smile vs non smile portraits)

INPUTS
<img src="https://github.com/user-attachments/assets/3a2d7feb-39e4-4019-b4b7-54891c80f59d" width="200" height="200">
<img src="https://github.com/user-attachments/assets/228087e1-ce4e-4c7c-9c19-5844d4bd3226" width="200" height="200">



OUTPUT
<img src="https://github.com/user-attachments/assets/f1bb8402-70ed-43be-a717-618505e3230c" width="200" height="200">


This is very cool as it doesn't require reverse diffusion so it's pretty fast, but again my initial idea is to start from a single input image and generate variations of it

**Reply by TimothyAlexisVass** [reply]
> my initial idea is to start from a single input image and generate variations of it

I'm not sure that I understand exactly. But if I do understand, then you can take the single input image, generate it to x% and then use that as input for your variations, just like you would with passing an x% done image from base to refiner, but use the base model instead.

**Reply by TimothyAlexisVass** [reply]
Here is an example, but notice that the syntax is for an old version of diffusers:
```
# Set the details for your model here:
import torch

from diffusers import StableDiffusionXLPipeline, StableDiffusionXLImg2ImgPipeline, AutoencoderKL, KDPM2AncestralDiscreteScheduler

vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16)
base = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0",
    vae=vae,
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True,
)
refiner = StableDiffusionXLImg2ImgPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-1.0",
    text_encoder_2=base.text_encoder_2,
    vae=vae,
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True,
)

base2 = StableDiffusionXLImg2ImgPipeline(**base.components)

base.scheduler = KDPM2AncestralDiscreteScheduler.from_config(base.scheduler.config, use_karras_sigmas=True)

_ = base.to("cuda")
_ = base2.to("cuda")
_ = refiner.to("cuda")

import os
import zipfile
import random
base.scheduler = base2.scheduler
prompt = f"Steampunk photo of antropomorphic (suricata), pixar, cute, in (gray) mecha armor looking happy, daylight lush nature park background"
quality = "intricate details even to the smallest particle, extreme detail of the enviroment, sharp portrait, well lit, interesting outfit, beautiful shadows, bright, photoquality, ultra realistic, masterpiece, 8k"
negative_prompt = "helmet, ugly, old, boring, photoshopped, tired, wrinkles, scar, gray hair, big forehead, crosseyed, dumb, stupid, cockeyed, disfigured, blurry, assymetrical, unrealistic, grayscale, black and white, bald, high hairline, balding, receeding hairline, grayscale, bad anatomy, unnatural irises, no pupils, blurry eyes, dark eyes, extra limbs, deformed, disfigured eyes, out of frame, no irises, assymetrical face, broken fingers, extra fingers, disfigured hands"
num_samples = 1
guidance_scale = 8
num_inference_steps = 30
height = 1024
width = 1024

prompt = prompt + ". " + quality

def cb(s,t,l):
    print(s,t)

# Set this to the folder you want to save the image to in Google Drive
output_dir = "drive/MyDrive/manually_generated"
os.makedirs(output_dir, exist_ok=True)

threshold = 0.3

latents = base(
    prompt,
    height=height,
    width=width,
    negative_prompt=negative_prompt,
    num_images_per_prompt=num_samples,
    num_inference_steps=num_inference_steps//2,
    guidance_scale=guidance_scale,
    generator=torch.manual_seed(2222),
    output_type = "latent",
    denoising_end=threshold
).images

colors = ["shining golden", "sparkling diamond", "off-white metallic", "radiant ruby", "fuschia"]
for color in colors:
    for animal in ["suricata", "giraffe", "tucan", "turtle"]:
        inference_prompt = prompt.replace("(gray)", color).replace("(suricata)", animal)
        print(inference_prompt)
        latents_2 = base2(
            inference_prompt,
            negative_prompt=negative_prompt,
            image=latents,
            num_inference_steps=num_inference_steps//2,
            guidance_scale=20,
            denoising_start=threshold,
            denoising_end=0.8,
            output_type = "latent"
        ).images
        refined = refiner(
            inference_prompt,
            negative_prompt=negative_prompt,
            image=latents_2,
            num_inference_steps=num_inference_steps,
            denoising_start=0.8
        )
        display(refined.images[0])
```
![image](https://github.com/user-attachments/assets/18be1423-fb1d-4b5c-8a9b-bef8bdd760b5)
![image](https://github.com/user-attachments/assets/09b22525-55d0-46fb-b616-a9db4925c02f)

**Reply by Clement-Lelievre** [reply]
> > my initial idea is to start from a single input image and generate variations of it
> 
> I'm not sure that I understand exactly. But if I do understand, then you can take the single input image, generate it to x% and then use that as input for your variations, just like you would with passing an x% done image from base to refiner, but use the base model instead.

my point is to avoid using the refiner or any call to the full model, for performance reasons.
I want to generate variations of an image just by sampling from the VAE and then decoding

the workflow, theoretically,  is:

1. use one input image
2. decode it into latents with the VAE
3. we're ineterested in the underlying proba distributions that the VAE inferred
4. sample other, close latents with the VAE
5. decode those latents into images; these images _should_ be close variants of the input image


But in this case it doesn't work, even when adding a lot of noise to the samples

**Reply by TimothyAlexisVass** [reply]
> But in this case it doesn't work, even when adding a lot of noise to the samples

I'm not quite sure why you think that should work?

The VAE just takes an image in encoded 4 channel latent space and decodes it to 3 channel (rgb) pixel space.

I don't see how doing it with the VAE would be any different than if you would do that with an image.

As far as I know, there is no "concept" (subject, form, emotion, etc) stored in the latent space, just a bunch of compressed 8x8 pixels really, so I don't expect that you would get get variations of that... but perhaps you have a better understanding and would correct me on that?

**Reply by Clement-Lelievre** [reply]
in the paper I reference above, as well as in my deep learning textbooks, they show that the VAE approximates the probability distributions that gave rise to the specific latents decoded. This is what I use above (the `DiagonalGaussianDistribution` in the diffusers code). This has the shape of the latents and contains a Normal distribution for each point in the latents. So for a 1024x1024 image and using SDXL's VAE you'd get 4x128x128 distributions. One latent is 4x128x128 samples, one from each of these distributions, IIUC. This is one of the reasons the VAE is more "powerful" than a "mere" autoencoder.

In my textbook (Deep learning with python by @fchollet, as well as in Deep Learning by Goodfellow, Courville and Bengio) they explain that if you sample from the VAE twice and decode both latents you should get different images.
I'm just not managing to get meaningfully different images in this case.

With SDXL's VAE, the stds are very very small relative to the mean of these distributions (like 1/10000th or so), I suspect this might be the reason why I don't get different images unless I put a lot of weight on the stds when sampling. I mean, sampling just falls really really close in the latent space due to the extremely small std.
But adding a lot of weight to the stds, in order to get far away from the initial latents, only results in producing artefacts in my experience.

**Reply by TimothyAlexisVass** [reply]
> in the paper I reference above, as well as in my deep learning textbooks, they show that the VAE approximates the probability distributions that gave rise to the specific latents decoded. 

Ok. I have no idea what you're talking about.
In my understanding, the latents tensor is just the image in a different, compressed format.

So I think you would need to show what you mean for me to understand.

Here is a simple example to show what I mean:
```
test = torch.flip(latents, dims=[0])

display(decode_latents(test)[0])
```
![image](https://github.com/user-attachments/assets/ed7780d3-9d33-434f-84d1-9315b97a461c)

```
test = torch.flip(latents, dims=[1])

display(decode_latents(test)[0])
```
![image](https://github.com/user-attachments/assets/68445742-bb72-4643-b4a5-88eca4c6625c)

```
test = torch.flip(latents, dims=[2])

display(decode_latents(test)[0])
```
![image](https://github.com/user-attachments/assets/503d5921-f6ae-48d1-a881-73152a3f16cd)

```
test = torch.flip(latents, dims=[3])

display(decode_latents(test)[0])
```
![image](https://github.com/user-attachments/assets/eddc1115-3532-4d6b-a2e1-ff6eb95c60b5)

**Reply by Clement-Lelievre** [reply]
sure, happy to show you, just let me know the best way to do this

**Reply by TimothyAlexisVass** [reply]
You _could_ write a blog post on Huggingface, that way a lot of people would learn something interesting. :)

Otherwise you could just post some images and examples here I guess.

**Reply by Clement-Lelievre** [reply]
I don’t really have examples or images to post, this is an idea and so far I didn’t manage to make it work.
The idea like I said above is: because the VAE encoding outputs normal distributions, let’s sample from these to get various latents of the same image, and then decode all of these latents in the hope that they will produce (slightly) different images.
It’d be an efficient way to get variations of an image. This has proven successful before, like in the paper I referenced. 

but so far in my xp, the means are so overwhelming relative to the stds, that no matter how much I perturb the samples, all the decoded latents look the same to the human eye.

**Reply by TimothyAlexisVass** [reply]
What I'm thinking is that... even if you manage to perturb the VAE latent space enough to produce visible variations in the output image, I am supposing that these variations are highly unlikely to be anything meaningful and will mostly manifest as introduced noise. I think that any variation you introduce by perturbing the latent vector is highly likely to affect random aspects of the reconstruction.

I think that:
VAE Latent Space is geared towards reconstruction and not structured for semantics. Like I wrote before, I don't think that the VAE has any idea of the concepts of the image.

The way I understand it is that VAEs are primarily designed to compress/reconstruct images between high-dimensional images/lower-dimensional latent space. I don't think either space has any inherent structure or meaning associated with specific semantic features of the image. While diffusion models can map particular latent dimensions to semantically meaningful variations, I don't think that standard VAEs do this. There might be something like that, but I don't think the SDXL is one of those.

The goal of a VAE is to reconstruct the input image with minimal error, not to generate diverse variations. I assume this leads to a tightly-packed latent space where different images are encoded very close to each other. In such a space, while the variations in the latent space may generally be smooth, again I don't think that they are tied to specific semantic features (such as changing a person's expression or altering the lighting in a scene).

**BUT!** It would be amazing if you would prove your idea, because it would indeed be an extremely cheap way to get a relatively vast amount of variation. Cheap enough that you could do this in real-time on a generated image and possibly solve things like wacky teeth and hands and all sorts of things... and _that_ would indeed be a breakthrough.
